`REFRESH` options for MVs -- Compute

ggevay opened the pull request:
This is the Compute part of the first part of the [REFRESH options for materialized views epic](https://github.com/MaterializeInc/materialize/issues/22878). It implements all the refresh options mentioned in the [design doc](https://github.com/ggevay/materialize/blob/refresh-every-design-doc/doc/developer/design/20231027_refresh_every_mv.md#solution-proposal), but it doesn't implement automatic replica management yet.

@jkosh44, could you please review the changes in `interval.rs`?

@teskje, could you please review `sequence_create_materialized_view` + the Compute parts + `refresh_schedule.rs`?

Note that in `sequence_create_materialized_view` I moved the timestamp selection outside of the `catalog_transact_with_side_effects`, as discussed with Matt [here](https://materializeinc.slack.com/archives/C02FWJ94HME/p1702488582281989?thread_ts=1702401378.106689&cid=C02FWJ94HME).

### Motivation

  * This PR adds a known-desirable feature: https://github.com/MaterializeInc/materialize/issues/22878

### Temporary Limitations

No automated replica management yet, as noted above.

If there is an index on the materialized view, then we have a problem after envd restarts, because the index's since will be bootstrapped to be 1 sec before the next refresh, which is most likely in the future, making queries on the MV block until the next refresh. We should enhance the frontier selection in the bootstrapping at some point, but for now I'll just tell the customer to please don't create indexes on these MVs.

This first version simply relies on Adapter read holds to keep the sinces close to the present moment. Eventually, we'll want to have custom compaction windows on `REFRESH EVERY` MVs (and their indexes), because relying on Adapter read holds precludes SERIALIZABLE reads from `REFRESH EVERY` MVs selecting somewhat older timestamps, making SERIALIZABLE reads not faster than STRICT SERIALIZABLE reads.

### Tips for reviewer

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
  - (First Nightly run for an old version of the PR: https://buildkite.com/materialize/nightlies/builds/5569)
  - (Second Nightly run for an old version of the PR: https://buildkite.com/materialize/nightlies/builds/5588 I think here only the "Legacy upgrade tests" failure is related to this PR.)
  - **Nightly run**: https://buildkite.com/materialize/nightlies/builds/5853
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

benesch commented on src/sql-parser/src/parser.rs:
```diff
@@ -4074,6 +4081,55 @@ impl<'a> Parser<'a> {
+            if self.parse_keyword(CREATION) {
+                Ok(WithOptionValue::Refresh(RefreshOptionValue::AtCreation))
+            } else if let Some(expr) = self.maybe_parse(Parser::parse_expr) {
+                Ok(WithOptionValue::Refresh(RefreshOptionValue::At(
+                    RefreshAtOptionValue { time: expr },
+                )))
+            } else {
+                Ok(WithOptionValue::Ident(ident!("at")))
+            }
+        } else if self.parse_keyword(EVERY) {
+            match self.maybe_parse(Parser::parse_value) {
+                Some(Value::String(interval)) => {
+                    let aligned_to = if self.parse_keywords(&[ALIGNED, TO]) {
+                        Some(self.parse_expr()?)
+                    } else {
+                        None
+                    };
+                    Ok(WithOptionValue::Refresh(RefreshOptionValue::Every(
+                        RefreshEveryOptionValue {
+                            interval,
+                            aligned_to,
+                        },
+                    )))
+                }
+                Some(v @ Value::Interval(_)) => {
+                    parser_err!(
+                            self,
+                            self.peek_prev_pos(),
+                            format!("Invalid value for REFRESH EVERY: `{v}`. The value should be a string parseable as an interval, e.g., '1 day'. The INTERVAL keyword should NOT be present!")
+                        )
+                }
+                Some(v) => {
+                    parser_err!(
+                            self,
+                            self.peek_prev_pos(),
+                            format!("Invalid value for REFRESH EVERY: `{v}`. The value should be a string parseable as an interval, e.g., '1 day'.")
+                        )
+                }
+                None => Ok(WithOptionValue::Ident(ident!("every"))),
+            }
```
This should not be in `parse_option_value`, since these special syntaxes are not in general valid option values. That's why you have the special cases around `at` and `on`. Instead, you should factor out a method like `fn parse_materialized_view_refresh_option_value`, and then in `parse_materialized_view_option` you should sniff out `MaterializedViewOptionName::Refresh` and call `parse_materialized_view_refresh_option_value` instead of `parse_optional_option_value`.

Look at `parse_kafka_broker` for a good example of how this works for an existing special cased option!

ggevay replied:
Ok, fixed!

benesch commented:
Exciting that you've gotten this working end to end, Gabor! This PR is at a size that's really hard to handle in GitHub reviews, though. Would recommend trying to split out the parsing, planning, and actual implementation into separate PRs. The parser stuff looks basically ready to go, for example, and could merge basically ASAP (just modulo the comments I dropped above). Totally okay for a parser-only PR to immediately error in the planner with "`REFRESH` not yet supported", and then for the next plannning-only PR to immediately error in the coordinator with "`REFRESH` not yet supported".

ggevay commented:
Ok, I'll split this into 3 PRs.

Thanks for the comments!

philip-stoev commented:
@ggevay is this ready for testing at this time?

ggevay commented:
Status: There are several minor issues that should be easy to fix, but there is one serious panic in slts. [I figured out what's causing the panic with Matt's help](https://materializeinc.slack.com/archives/C02FWJ94HME/p1702403728460379?thread_ts=1702401378.106689&cid=C02FWJ94HME), but it's not entirely clear how to fix it. Working on it.

ggevay commented on src/sql/src/names.rs:
```diff
@@ -1783,6 +1784,22 @@ impl<'a> Fold<Raw, Aug> for NameResolver<'a> {
+                    interval,
+                    aligned_to: aligned_to.map(|e| self.fold_expr(e)),
+                })),
+            },
```
Ah yes, fixed!

ggevay commented:
I've fixed all known issues, except for the "Legacy upgrade tests" fail in the Nightly run. (The normal upgrade tests are passing after I added the migration, as discussed with Nikhil [here](https://materializeinc.slack.com/archives/C063H5S7NKE/p1702390621759829?thread_ts=1699543250.405409&cid=C063H5S7NKE).)

ggevay commented:
I've addressed @benesch's comments.

I've split the PR into 2 PRs in the end: parsing+planning is one PR, and Compute is another. I didn't want to split the parsing from the planning because they are quite closely tied, so it probably makes sense if the same person reviews both the parsing and planning, and at the same time. Please let me know if it's ok like this.

The other PR is https://github.com/MaterializeInc/materialize/pull/23870.
And the second PR is this one, which includes also the commits from the other one.

This is ready for review now! It would be great if we could make it into the release cut.

ggevay commented:
@philip-stoev, this is ready for testing. (Nightly has several failures, which are either minor or unrelated, and are being discussed [here](https://materializeinc.slack.com/archives/C01LKF361MZ/p1702498564721889).)

jkosh44 commented on src/repr/src/adt/interval.rs:
```diff
@@ -376,6 +377,25 @@ impl Interval {
 
+    /// Computes the total number of milliseconds in the interval. Discards fractional millisecond!
+    pub fn as_milliseconds(&self) -> i128 {
+        // unwrap is safe because i32::MAX/i32::MIN number of months will not overflow an i128 when
+        // converted to milliseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Month,
+            DateTimeField::Milliseconds,
+            i128::from(self.months),
+        ).unwrap() +
+        // unwrap is safe because i32::MAX/i32::MIN number of days will not overflow an i128 when
+        // converted to milliseconds.
+        Self::convert_date_time_unit(
+            DateTimeField::Day,
+            DateTimeField::Milliseconds,
+            i128::from(self.days),
+        ).unwrap() +
+        i128::from(self.micros) / 1000
```
Could this all be replaced with 
```Rust
self.as_microseconds() / 1000
```
Or 
```Rust
Self::convert_date_time_unit(
            DateTimeField::Microseconds,
            DateTimeField::Milliseconds,
            self.as_microseconds(),
).unwrap()
```
?

ggevay replied:
Done: replaced with
```
self.as_microseconds() / 1000
```

jkosh44 commented:
> @jkosh44, could you please review the changes in interval.rs?

Do you still want me to only look at this file or should I look at all the adapter changes?

benesch commented:
> I didn't want to split the parsing from the planning because they are quite closely tied, so it probably makes sense if the same person reviews both the parsing and planning, and at the same time. Please let me know if it's ok like this.

Not worth the trouble to split any further now, but in the future I would encourage you to split out the parser PRs and send them up early and often! Parser PRs that are <500 lines are an absolute delight to review; a PR with parsing and planning at 1k lines is feasible but much harder. I'd much rather have a steady stream of small, incremental PRs to review than one or two big PRs. Even if it means that the net number of PRs reviewed is higher, because later PRs have to make small tweaks to the first PRs, the reviews are much higher quality when there are less lines changed. Also, as the PR author, I find it relieving to be able to peel off layers and get them merged on to main; each merged PR is one less possibility for a merge conflict on a rebase.

teskje commented on src/compute-types/src/sinks.rs:
```diff
@@ -75,6 +95,7 @@ impl RustType<ProtoComputeSinkDesc> for ComputeSinkDesc<CollectionMetadata, mz_r
             with_snapshot: self.with_snapshot,
             up_to: Some(self.up_to.into_proto()),
             non_null_assertions: self.non_null_assertions.into_proto(),
+            refresh_schedule: self.refresh_schedule.as_ref().map(|r| r.into_proto()),
```
I think this can be just:
```rust
            refresh_schedule: self.refresh_schedule.into_proto(),
```

ggevay replied:
Yes, thx

ggevay replied:
Yes, thx

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -58,7 +62,13 @@ where
-        let desired_collection = sinked_collection.map(Ok).concat(&err_collection.map(Err));
+        let mut desired_collection = sinked_collection.map(Ok).concat(&err_collection.map(Err));
+
+        // If `REFRESH EVERY` was specified, round up timestamps.
```
Some of the comments in this commit explicitly mention `REFRESH EVERY`, but there are other `REFRESH` options to which they apply as well, right?

ggevay replied:
Right!

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+///
+/// Note that this currently only works with 1-dim timestamps. (This is not an issue for WMR,
+/// because iteration numbers should disappear by the time the data gets to the Persist sink.)
+fn round_up<G>(
```
Can we rename the operator (both the function name here and the name we pass to `OperatorBuilder::new` below) to something that makes it more obvious what it's doing? I propose `apply_refresh`.

ggevay replied:
Done

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let mut builder = OperatorBuilder::new("round_up".to_string(), coll.scope().clone());
```
The `.clone()` here is not needed.

ggevay replied:
thx

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+{
+    let mut builder = OperatorBuilder::new("round_up".to_string(), coll.scope().clone());
+    let (mut output_buf, output_stream) = builder.new_output();
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
```
Since inputs with an empty path summary are non-standard and a potential source for bugs/confusion, could you add a comment explaining why we need to disconnect the reachability graph here?

ggevay replied:
Done

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+    let (mut output_buf, output_stream) = builder.new_output();
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
+    builder.build(move |capabilities| {
+        let mut capability = Some(capabilities.into_element());
```
Maybe also add a comment here to explain that:
* This capability directly controls this operators output frontier (because we have disconnected the input above).
* We wrap it in an `Option` so we can drop it to advance to the empty output frontier when we know no refresh will happen again.
* We must be careful that we only ever emit output updates at times that are at or beyond this capability.

ggevay replied:
Done

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
```
I think you need to move the capability check into the `input.for_each`. Even if we will never produce output anymore, we still have to drain the operator inputs, otherwise the data that's pushed there from upstream will remain in the input channel and leak.

ggevay replied:
Hmm, good catch!

But actually, I'd like to just terminate the dataflow when we'll never produce output anymore (i.e., when we are past the last ever refresh). I was hoping dropping the capability would already have this effect. Isn't this the case? Or what is the standard way to signal from inside a dataflow that I'm completely done forever?

teskje replied:
We don't have a way to communicate this from inside the dataflow. Just dropping the capability means the write frontier will advance to the empty frontier, which communicates that the dataflow is not going to write any more data. But upstream operators don't care (or know) about this and continue to do their thing. The `ComputeState` doesn't drop dataflows when they advance to the empty frontier. It only drops them when it receives an `AllowCompaction([])` command from the controller. But the controller doesn't allow compaction to the empty frontier in response to dataflows advancing to the empty frontier, because others might still want to read the produced data.

The current behavior makes sense for indexes, but not necessarily for MVs. I think we could drop MV dataflows immediately when they have advanced to the empty frontier, since we don't need the dataflow to read the data from persist. But we don't have this optimization at the moment because it wasn't useful.

antiguru replied:
+1 to what @teskje said. The operator must not stop consuming its inputs when it doesn't want to produce output anymore, otherwise we'll have a memory (and CPU) leak.

ggevay replied:
Done. Now I've also set the `until` to the last refresh, so we shouldn't ever receive data after we drop our capability when the last refresh is complete.

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
```
Here it would also be nice to have a comment stating that we can't use `_cap` to get an output session because we might have advanced our output frontier already beyond the frontier of this capability.

antiguru replied:
Good point! I think it's not trivially clear that this operator is correct in all corner cases. Specifically, it relies on properties that `round_up_timestamp` is deterministic, which is not documented as such.

ggevay replied:
Thanks, I've added the comment, and I've also documented that `round_up_timestamp` is monotonic, which I think is the needed property.
```
// Round up the frontier.
// Note that `round_up_timestamp` is monotonic. This is needed to ensure that the timestamp (t) of any
// received data that has a larger timestamp than the original frontier (f) will get rounded up to a time
// that is at least at the rounded up frontier. In other words, monotonicity ensures that
// when `t >= f` then `round_up_timestamp(t) >= round_up_timestamp(f)`.
```

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1191,87 @@ where
+        }
+    });
+
+    use differential_dataflow::AsCollection;
```
Nit: I'd prefer importing this trait once rather than at every call site.

ggevay replied:
done

teskje commented on src/repr/src/timestamp.rs:
```diff
@@ -182,6 +198,54 @@ impl Timestamp {
+            interval,
+            aligned_to,
+        }: &RefreshEvery,
+    ) -> Option<Self> {
```
I noticed that the method to round up a timestamp according to a `RefreshSchedule` is defined on `RefreshSchedule`, but the method to round up a timestamp according to a `RefreshEvery` is defined on `Timestamp`. Could we move this logic to a `RefreshEvery::round_up_timestamp` method for consistency?

ggevay replied:
Done

teskje commented on src/repr/src/timestamp.rs:
```diff
@@ -182,6 +198,54 @@ impl Timestamp {
+        // Planning ensured that
+        // - interval.months == 0, so we don't need to deal with months being of variable size.
+        // - The interval can be max 27 days, so the cast to u64 won't overflow.
+        // - The interval is positive, so the cast to u64 won't underflow.
```
Instead of relying on specific planning behavior here, could you just document these cases in the `# Panics` section of this method and then let the planner worry about this methods requirements?

teskje replied:
Another requirement is that the interval (in milliseconds) must not be 0, because otherwise we'll get a division by zero panic.

ggevay replied:
Done

teskje commented on src/repr/src/timestamp.rs:
```diff
@@ -182,6 +198,54 @@ impl Timestamp {
+        };
+        assert!(result.internal >= self.internal);
+        assert!(result.internal - self.internal <= interval);
+        Some(result)
```
If we are worried about the performance of this in the hot persist_sink path, how about making these `assert`s (non-logging) soft-asserts?

ggevay replied:
For now, I'd sleep better if these are always-on asserts. They should be negligible compared to all the other things  going on in this function (or, not to mention, in the entire system). E.g. 64-bit integer division is around 15 clock cycles on modern Intel CPUs.

(An even bigger thing is the `interval.as_milliseconds()` call at the beginning of the function, which involves several integer divisions. I think I'll move the `as_milliseconds()` call into the creation of `RefreshEvery` from this function.)

teskje replied:
Fine by me! In this case it would be good to have a `TODO` comment to downgrade the asserts once we have built some confidence in the code.

ggevay replied:
Ok, done

teskje reviewed (changes requested):
Many nits, but also two important things:
* We need to make sure the `as_of` is also selected correctly during bootstrapping.
* The refresh operator needs to drain its inputs even when it won't produce output anymore.

ggevay commented on src/compute-types/src/sinks.rs:
```diff
@@ -21,27 +23,43 @@ include!(concat!(env!("OUT_DIR"), "/mz_compute_types.sinks.rs"));
     type Parameters = ();
 
     fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+        let interval_strategy: BoxedStrategy<RefreshSchedule> = (
```
Yes

ggevay commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1054,6 +1049,28 @@ impl Coordinator {
 
+        // Timestamp selection
+        let as_of = {
+            // Normally, `since` should be the least_valid_read.
```
Yes, sorry!

ggevay commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1054,6 +1049,28 @@ impl Coordinator {
+            // `since` to the first refresh. This is because we'd like queries on the MV to
+            // block until the first refresh (rather than to show an empty MV).
+            if let Some(refresh_schedule) = refresh_schedule {
+                if let Some(ts) = as_of.as_option() {
```
Yes, good idea to sanity check this!

ggevay commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1054,6 +1049,28 @@ impl Coordinator {
+                }
+            }
+            as_of
+        };
```
Oopsie, good catch! I'll think this through carefully.

Btw. I'm also thinking that I should add a test that actually does a restart of envd. (My td test only restarts the replica.) Although even such a test wouldn't have caught this bug, because the bug would happen only if a crash happens at a very specific point in time.

teskje replied:
You might be able to trigger this today by installing the MV on a cluster with no replicas. I think the storage controller initializes the shard with an `upper = [0]` and it is only advanced from that when compute installs the `persist_sink` and advances the frontier... which it can't do without a replica.

ggevay replied:
I think I fixed this in 0a96a639314db78b5d3b08c6a69ee44fab945c7f, but I'll think this through tomorrow again.

ggevay replied:
Discussed with @teskje on Slack.

ggevay commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1054,6 +1049,28 @@ impl Coordinator {
+            // block until the first refresh (rather than to show an empty MV).
+            if let Some(refresh_schedule) = refresh_schedule {
+                if let Some(ts) = as_of.as_option() {
+                    let mut rounded_up_as_of = Antichain::new();
+                    if let Some(rounded_up_ts) = refresh_schedule.round_up_timestamp(*ts) {
+                        rounded_up_as_of.insert(rounded_up_ts);
+                    } else {
+                        return Err(AdapterError::MaterializedViewWouldNeverRefresh(name.item));
+                    }
+                    as_of = rounded_up_as_of;
```
Yes, thanks!

antiguru commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1190,86 @@ where
+                    output_buf
+                        .activate()
+                        .session(&capability)
+                        .give_container(&mut buffer);
```
`give_iterator` if the buffer might be not 100% filled.

ggevay replied:
I did some refactoring and now there is a `consolidate_updates` and then a `shrink_to_fit` on the `buffer` (after the `retain_mut`). So `give_container` is ok in this case, right?

antiguru replied:
Instead of consolidating here you might be better off using the consolidate buffer! https://github.com/MaterializeInc/materialize/blob/3ada8fb079c5e6b36e2cf1c5b0de1a296f3310a6/src/timely-util/src/buffer.rs#L41

ggevay replied:
Ok, I switched to `ConsolidateBuffer`, could you please check? I added a method to `ConsolidateBuffer`.

255ec4bb87da544f318792b93f3cd79cb7a5bd76

ggevay replied:
Discussed on Slack.

antiguru commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1190,86 @@ where
     (output_stream, token)
 }
+
+/// This is for REFRESH options on materialized views. It adds an operator that rounds up the
```
My gut feeling is that this operator is too complex and does multiple things at once, which we might want to separate out. It rounds up the timestamp (this could be a simple `map`), it delays the data (there's the `delay` operator), and it eventually ceases producing data. I think strongly that it should be factored in this way to make it clear what the operator is doing, and to avoid the pitfalls it currently has.

I think the property that we'd want to close this sink at some point needs to be encoded in the `up_to` of the dataflow itself, and not here.

teskje replied:
Correct me if I'm wrong, but I don't think the `delay` operator can do what we need to do here. Note that we don't just need to delay the data, but we also have to make sure that the operators frontier advances to the next refresh interval as soon as it has produced all outputs for the current interval. This is important because otherwise the MV wouldn't remain queryable (under strict serializable) while we take the replica down between refresh intervals.

AIUI, to make the operators frontier jump to the future, you need to fiddle with the input path summaries somehow. The `delay` operator doesn't do that, so it's output frontier will always be bound by the capabilities still held by upstream operators, which would prevent it from jumping into the future.

ggevay replied:
We'll discuss this next week.

Btw. this is the relevant part of the design doc: https://github.com/ggevay/materialize/blob/refresh-every-design-doc/doc/developer/design/20231027_refresh_every_mv.md#b-rounding-up-timestamps-going-into-the-persist-sink

antiguru replied:
Yup, the delay operator doesn't have any special connectivity, so we'd need to roll our own.

ggevay replied:
I've refactored the operator as discussed in Timely office hours.

antiguru commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -58,7 +62,13 @@ where
-        let desired_collection = sinked_collection.map(Ok).concat(&err_collection.map(Err));
+        let mut desired_collection = sinked_collection.map(Ok).concat(&err_collection.map(Err));
+
+        // If any non-trivial `REFRESH` options were specified, round up timestamps.
```
We should separate the SQL-level concept (`REFRESH`) from the lir-level concept and make it such that it's clear on its own.

ggevay replied:
I've changed the comment.

antiguru commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -46,6 +49,7 @@ impl<G> SinkRender<G> for PersistSinkConnection<CollectionMetadata>
 where
     G: Scope<Timestamp = Timestamp>,
 {
+    #[allow(clippy::as_conversions)]
```
Stray? In any case, let's try to avoid this lint.

antiguru commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1181,3 +1190,86 @@ where
+                    let mut cached_ts: Option<Timestamp> = None;
+                    let mut cached_rounded_up_ts = None;
+                    buffer
+                        .drain_filter_swapping(|(_d, ts, _r)| {
```
I think you can just use [`retain_mut`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.retain_mut)

ggevay replied:
Yes, thanks!

antiguru commented on src/repr/src/adt/interval.rs:
```diff
@@ -376,6 +377,25 @@ impl Interval {
         i128::from(self.micros)
     }
 
+    /// Computes the total number of milliseconds in the interval. Discards fractional millisecond!
```
```suggestion
    /// Computes the total number of milliseconds in the interval. Discards fractional milliseconds!
```

ggevay replied:
fixed

antiguru reviewed (changes requested):
Snapshot of my comments thus far, I'll continue later.

ggevay commented:
> in the future I would encourage you to split out the parser PRs and send them up early and often!

Ok, I see what you mean; I've just resolved merge conflicts in 14 files :facepalm:

aalexandrov commented on src/adapter/src/coord/sequencer/inner.rs:
```diff
@@ -1036,10 +1039,33 @@ impl Coordinator {
+                    let Some(rounded_up_ts) = refresh_schedule.round_up_timestamp(*ts) else {
+                        return Err(AdapterError::MaterializedViewWouldNeverRefresh(name.item));
+                    };
+                    as_of = Antichain::from_elem(rounded_up_ts);
```
I suggest to add some a structured log line with some message and `rounded_up_ts` and `refresh_schedule` reported as key-value pairs.

def- reviewed (changes requested):
This seems to break legacy upgrade tests:
https://buildkite.com/materialize/nightlies/builds/5670#018c87b0-5e8f-404a-83d5-317b66c4be98
```
--- check-from-v0.27.0-subquery.td
^^^ +++
check-from-v0.27.0-special-functions.td:16:1: error: non-matching rows: expected:
[["materialize.public.special_functions_view", "CREATE MATERIALIZED VIEW \"materialize\".\"public\".\"special_functions_view\" IN CLUSTER \"default\" AS SELECT * FROM \"materialize\".\"public\".\"special_functions\" WHERE \"mz_catalog\".\"mz_now\"() > \"f1\""]]
got:
[["materialize.public.special_functions_view", "CREATE MATERIALIZED VIEW \"materialize\".\"public\".\"special_functions_view\" IN CLUSTER \"default\" WITH (REFRESH = ON COMMIT) AS SELECT * FROM \"materialize\".\"public\".\"special_functions\" WHERE \"mz_catalog\".\"mz_now\"() > \"f1\""]]
Poor diff:
- materialize.public.special_functions_view "CREATE MATERIALIZED VIEW \"materialize\".\"public\".\"special_functions_view\" IN CLUSTER \"default\" AS SELECT * FROM \"materialize\".\"public\".\"special_functions\" WHERE \"mz_catalog\".\"mz_now\"() > \"f1\""
+ materialize.public.special_functions_view "CREATE MATERIALIZED VIEW \"materialize\".\"public\".\"special_functions_view\" IN CLUSTER \"default\" WITH (REFRESH = ON COMMIT) AS SELECT * FROM \"materialize\".\"public\".\"special_functions\" WHERE \"mz_catalog\".\"mz_now\"() > \"f1\""
```
```

def- commented on src/compute-types/src/sinks.rs:
```diff
@@ -21,27 +23,43 @@ include!(concat!(env!("OUT_DIR"), "/mz_compute_types.sinks.rs"));
+                }],
+                ats,
+            })
+            .boxed();
```
You are right, coverage flagged this as uncovered! The code certainly looks interesting, so I think this should be tested.

ggevay replied:
Fixed in b49aef07b9aec502dcadeee8d7c94609c19dc833

def- reviewed (changes requested):
I found 2 separate panics with my [one-off draft](https://github.com/MaterializeInc/materialize/pull/24210) that hackily converts all MVs to `REFRESH EVERY '2 seconds'`:
```
thread 'timely:work-0' panicked at /cargo/git/checkouts/timely-dataflow-70b80d81d6cabd62/de20aa8/timely/src/dataflow/operators/capability.rs:139:13: Attempted to downgrade Capability { time: 1704332460774, internal: "..." } to 774, which is not beyond the capability's time.
```
and
```
thread 'coordinator' panicked at src/sql/src/plan/statement/ddl.rs:2148:29: ALIGNED TO should have been filled in by purification
```
I'm guessing those are both unexpected. The non-matching rows mostly looked like they are not wrong results, but I'm not sure about the controller-frontiers:
```
controller-frontiers.td:305:1: error: non-matching rows: expected:
[["0", "<null>"]]
got:
[["751", "<null>"]]
Poor diff:
- 0 <null>
+ 751 <null>
```
Test results are in https://buildkite.com/materialize/tests/builds/72256 and https://buildkite.com/materialize/nightlies/builds/5842

ggevay commented:
> I found 2 separate panics with my https://github.com/MaterializeInc/materialize/pull/24210 that hackily converts all MVs to REFRESH EVERY '2 seconds':

@def- thank you very much for this test!

I've fixed both panics:
- The `Attempted to downgrade Capability` panic was fortunately already fixed yesterday evening (caught by @antiguru from looking at the code) by https://github.com/MaterializeInc/materialize/pull/23819/commits/56d8e123ce89b8c75c0894db79eb7e19b2fd4fbc
- The `ALIGNED TO` panic was caused by forgetting to purify `EXPLAIN CREATE MATERIALIZED VIEW`, fixed by 4e9f0701cd6230d2b6956e174ef82cd8785a2ec6

I pushed the fixes to your PR, and running CI again. Hopefully, there will be no panics now.

The `non-matching rows` might be ok, but I'll definitely investigate them.

def- reviewed (commented):
Is it expected that the `REFRESH` option can be repeated? Only the last will have an effect, but the first one is still evaluated, so this panics:
```sql
CREATE SOURCE counter        
  FROM LOAD GENERATOR COUNTER
  (TICK INTERVAL '1s');      

CREATE MATERIALIZED VIEW mv8 WITH (REFRESH AT mz_unsafe.mz_panic('heh'), REFRESH EVERY '1 second') AS SELECT * FROM counter;
```
Here I'm only seeing the refresh every 10 seconds, not at 2 seconds after the current time:
```
CREATE MATERIALIZED VIEW mv7 WITH (REFRESH AT mz_now()::text::int8 + 2000, REFRESH EVERY '10 seconds') AS SELECT * FROM counter;     COPY (SUBSCRIBE (SELECT * FROM mv8)) TO STDOUT;
```
I'm not sure if this is how we handle options in general or if we want to block this off since someone could accidentally use it and expect both refreshes to have an effect.

ggevay commented:
@def- 

The first one is expected to panic, because both REFRESH options are evaluated, so the `mz_unsafe.mz_panic` will be called.

The second one creates `mv7`, but then reads from `mv8`. If I make the SUBSCRIBE read from `mv7`, then the refresh that is 2 seconds after the creation does happen for me. However, there is actually a bug here: I treated the last REFRESH AT to be the last refresh ever, so the SUBSCRIBE terminates, when actually it should go on indefinitely. I'll fix this in a moment.

teskje commented on src/adapter/src/coord.rs:
```diff
@@ -1558,9 +1559,28 @@ impl Coordinator {
                     df_desc.set_as_of(as_of);
 
+                    // If we have a refresh schedule that has a last refresh, then set the `until` to the last refresh.
+                    let until = mview
+                        .refresh_schedule
+                        .as_ref()
+                        .map(|refresh_schedule| {
+                            refresh_schedule
+                                .last_refresh()
+                                .map(|last_refresh| last_refresh.try_step_forward())
+                        })
+                        .flatten()
+                        .flatten();
```
I think this can be simplified:
```suggestion
                                        let until = mview
                        .refresh_schedule
                        .as_ref()
                        .and_then(|s| s.last_refresh())
                        .and_then(|r| r.try_step_forward());
```

ggevay replied:
Ah, yes, thanks! Fixed in 1f4b5af3b8792d5bf05230d20dcd04a70dce827b

teskje commented on src/adapter/src/coord/sequencer/inner/create_materialized_view.rs:
```diff
@@ -285,6 +281,40 @@ impl Coordinator {
+        // If we have a refresh schedule that has a last refresh, then set the `until` to the last refresh.
+        // (If the `try_step_forward` fails, then no need to set an `until`, because it's not possible to get any data
+        // beyond that last refresh time, because there are no times beyond that time.)
+        let until = refresh_schedule
+            .map(|refresh_schedule| {
+                refresh_schedule
+                    .last_refresh()
+                    .map(|last_refresh| last_refresh.try_step_forward())
+            })
+            .flatten()
+            .flatten();
```
Same here:

```suggestion
        let until = refresh_schedule
            .and_then(|s| s.last_refresh())
            .and_then(|r| r.try_step_forward());
```

ggevay replied:
Fixed in 1f4b5af3b8792d5bf05230d20dcd04a70dce827b

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1209,3 +1218,114 @@ where
+                // data that is after the last refresh by setting the `until` of the dataflow to the last refresh.
+                let capability = capability
+                    .as_mut()
+                    .expect("should have a capability if we received data");
```
How certain are we about this part? It sounds reasonable, on first glance. But we only provide the `until` to index and source import operators, and we don't have anything in place that prevents other upstream operators from inventing updates from times in the future, right?

I would feel better if we handled this gracefully by soft-asserting and returning instead.

ggevay replied:
One thing that we briefly checked is that the rendering of temporal filters seems to be also mindful of the `until`: if it were to delay some addition or retraction to after the `until`, then it drops those instead.

ggevay replied:
We had a meeting with @teskje and @antiguru and decided to handle this and similar cases with soft panics in this operator.

teskje commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -1209,3 +1218,114 @@ where
+                        None => {
+                            // This record is after the last refresh, which is not possible because we set the dataflow
+                            // `until` to the last refresh.
+                            unreachable!("Received data after the last refresh");
```
Same here, can we handle this error without crashing?

ggevay replied:
Changed to `soft_panic_or_log`

teskje reviewed (approved):
I still have some questions about what setting `until` actually guarantees about the times that arrive at a dataflow output, and I'd like to see a test for the bootstrap as-of selection. But given that this will be behind a feature flag and disabled by default, I'm fine with merging this version.

def- reviewed (changes requested):
Thanks, not seeing any panics anymore

> The non-matching rows might be ok, but I'll definitely investigate them.

I took a quick look based on the lastest runs in https://buildkite.com/materialize/tests/builds/72270 and https://buildkite.com/materialize/nightlies/builds/5848
These look worrying:

```
controller-frontiers.td:305:1: error: non-matching rows: expected:
[["0", "<null>"]]
got:
[["603", "<null>"]]
Poor diff:
- 0 <null>
+ 603 <null>
```

```
kafka-avro-upsert-sinks.td:186:1: error: missing record 0: Record {
    headers: [
        "1",
    ],
    key: Some(
        {
            (
                "a",
                Long(1),
            ),
        },
    ),
    value: Some(
        {
            (
                "a",
                Long(1),
            ),
            (
                "b",
                Long(1),
            ),
        },
    ),
}
```

```
temporal.td:18:1: error: executing query failed: db error: ERROR: Timestamp (0) is not valid for all inputs: [Antichain { elements: [700] }]: ERROR: Timestamp (0) is not valid for all inputs: [Antichain { elements: [700] }]
     |
  17 |
  18 | > SELECT * FROM one_bound1 AS OF 0
     | ^
```

```
> SELECT * FROM foo;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 22s
^^^ +++
timestamps-debezium-kafka.td:160:1: error: non-matching rows: expected:
[["1", "1"], ["2", "2"]]
got:
[]
Poor diff:
- 1 1
- 2 2
```

```
> SELECT
  frontiers.read_frontier,
  frontiers.write_frontier
FROM mz_internal.mz_frontiers frontiers
JOIN mz_materialized_views mvs
  ON frontiers.object_id = mvs.id
WHERE
  mvs.name = 'mv2'
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 22s
^^^ +++
controller-frontiers.td:305:1: error: non-matching rows: expected:
[["0", "<null>"]]
got:
[["6", "<null>"]]
Poor diff:
- 0 <null>
+ 6 <null>
```

ggevay commented:
@def- 

> These look worrying:

Thanks, I'll check them!

Btw. we just now did some changes with @antiguru and @teskje, so I'll rebase the testing PR now to re-run. Saving the links to the old run that your above comment refers to:
- https://buildkite.com/materialize/tests/builds/72270
- Nightly: https://buildkite.com/materialize/nightlies/builds/5848

ggevay commented:
@def- The non-matching rows where `mz_internal.mz_frontiers` is involved is expected, because REFRESH options modify how the frontiers move.

ggevay commented:
@def- The

`temporal.td:18:1: error: executing query failed: db error: ERROR: Timestamp (0) is not valid for all inputs:`

is also expected: REFRESH options modify what times are valid to read from: originally 0 was valid because this MV has only constant inputs, but with the REFRESH EVERY the first valid read time will be the time of the first refresh, which will be `mz_now()` by default.

ggevay commented:
@def- 
`kafka-avro-upsert-sinks.td`: The non-matching rows here are also expected: The distribution of data among those 3 `kafka-verify-data` calls is based on data timestamps in the MV, which is modified by the REFRESH options (times are rounded up to the nearest refresh).

ggevay commented:
@def- 
`timestamps-debezium-kafka.td`: This one is tricky, but also expected: This test is manually controlling the progress of the Kafka topics. A REFRESH EVERY mat view is reading from such a topic, and it gets stuck, because the topic never makes enough progress to reach the time of the first refresh. (Td waiting for the rows to match up is not helping, because wall-clock time is not helping the progress of the Kafka topic.)

ggevay commented:
@def- 
I went through all the "These look worrying:". Thanks for flagging them!

antiguru reviewed (approved):
I think this is fine as-is. There are several places that could panic, but all seem they could only be triggered when something upstreams fails, which isn't ideal, but something we could revisit later.

I did not review adapter and testing changes.

ggevay commented:
Thanks @antiguru!

I'm making some cosmetic changes, and then run Nightly and then merge it.

We'll address @def-'s coverage concerns in a follow-up PR that will add lots of new tests.

ggevay commented:
Merged! Thank you very much for all the discussions and reviews!
